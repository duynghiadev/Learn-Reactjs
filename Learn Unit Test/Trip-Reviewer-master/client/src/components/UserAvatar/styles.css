.fit-parent {
  width: 100%;
  height: 100%;
}
.user-avatar-container .user-avatar-spinner {
  line-height: 200px;
  text-align: center;
  vertical-align: middle;
}
.user-avatar-container .avatar-picture img {
  box-shadow: 1px 3px 15px black;
  border: 1px solid;
  object-fit: cover;
}
