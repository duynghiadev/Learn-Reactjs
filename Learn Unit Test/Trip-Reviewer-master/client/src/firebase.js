import * as firebase from 'firebase';

const config = {
  apiKey: 'AIzaSyBLEX7iZntFl-xVHhe9CRRRQiAxHnW1dWg',
  authDomain: 'trip-reviewer-2018.firebaseapp.com',
  databaseURL: 'https://trip-reviewer-2018.firebaseio.com',
  projectId: 'trip-reviewer-2018',
  storageBucket: 'trip-reviewer-2018.appspot.com',
  messagingSenderId: '721559559011',
};

firebase.initializeApp(config);

export { firebase };
