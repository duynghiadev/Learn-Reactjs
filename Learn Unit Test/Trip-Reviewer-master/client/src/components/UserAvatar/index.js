import React, { PureComponent } from 'react';
import Loader from '../Loaders/PulseLoader';

import { firebase } from '../../firebase';
import './styles.css';

class UserAvatar extends PureComponent {
  state = {
    src: '',
    isLoading: true,
  };

  componentWillMount() {
    this.oldFilename = '';
    this.updateImage(this.props);
  }

  componentDidMount() {
    this._isMounted = true;
  }

  componentWillReceiveProps(nextProps) {
    this.updateImage(nextProps);
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  updateImage(props) {
    const { filename } = props;

    if (filename && filename !== this.oldFilename) {
      firebase
        .storage()
        .ref('avatars')
        .child(filename)
        .getDownloadURL()
        .then(this.handleGetUrlSuccess);

      this.oldFilename = filename;
    } else if (!filename) {
      this.setState({ src: '/images/avatar.png', isLoading: false });
    }
  }

  handleGetUrlSuccess = url => {
    if (this._isMounted) {
      this.setState({ src: url, isLoading: false });
    }
  };

  renderWithLoading() {
    const { isLoading, src } = this.state;

    if (isLoading) {
      return <Loader className="user-avatar-spinner" color="grey" size="16px" margin="4px" />;
    }
    return (
      <div className="avatar-picture fit-parent">
        <img className="fit-parent" alt="avatar" src={src} />
      </div>
    );
  }

  render() {
    return <div className="user-avatar-container fit-parent">{this.renderWithLoading()}</div>;
  }
}

export default UserAvatar;
