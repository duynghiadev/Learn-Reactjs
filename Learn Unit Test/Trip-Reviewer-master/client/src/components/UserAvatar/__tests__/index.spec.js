import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';

import UserAvatar from '..';
import { firebase } from '../../../firebase';

jest.mock('./../../Loaders/PulseLoader', () => 'Loader');
jest.mock('./../../../firebase', () => ({
  firebase: {
    storage: jest.fn(() => ({
      ref: jest.fn(() => ({
        child: jest.fn(() => ({
          getDownloadURL: jest.fn(() => ({
            then: jest.fn(),
          })),
        })),
      })),
    })),
  },
}));

const mockComponent = props => <UserAvatar {...props} />;

describe('<UserAvatar />', () => {
  it('should render component with no filename', () => {
    const props = {
      filename: '',
    };
    const tree = create(mockComponent(props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render component with filename', () => {
    const props = {
      filename: 'filename',
    };
    const tree = create(mockComponent(props)).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should change state when handleGetUrlSuccess is called', () => {
    const instance = shallow(mockComponent({})).instance();

    instance.handleGetUrlSuccess('url');

    expect(instance.state.src).toEqual('url');
    expect(instance.state.isLoading).toEqual(false);
  });

  it('should not change state when handleGetUrlSuccess is called for unmounted', () => {
    const instance = shallow(mockComponent({})).instance();

    instance.componentWillUnmount();
    instance.handleGetUrlSuccess('url');

    expect(instance.state.src).not.toEqual('url');
  });

  it('should call updateImage when componentWillReceiveProps is called', () => {
    const updateImage = jest.fn();
    const instance = shallow(mockComponent({})).instance();

    instance.updateImage = updateImage;
    instance.componentWillReceiveProps('test');

    expect(updateImage).toHaveBeenCalledWith('test');
  });

  it('should not call firebase.storage when updateImage is called with the same filename', () => {
    const filename = 'filename';
    const instance = shallow(mockComponent({})).instance();

    instance.oldFilename = filename;
    instance.updateImage({ filename });

    expect(firebase.storage).toHaveBeenCalledTimes(1);
  });

  it('should reset this._isMounted when componentWillUnmount is called', () => {
    const instance = shallow(mockComponent({})).instance();

    instance.componentWillUnmount();

    expect(instance._isMounted).toEqual(false);
  });
});
