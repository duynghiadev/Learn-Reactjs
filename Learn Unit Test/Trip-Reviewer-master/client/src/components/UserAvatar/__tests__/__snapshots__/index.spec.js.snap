// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAvatar /> should render component with filename 1`] = `
<div
  className="user-avatar-container fit-parent"
>
  <Loader
    className="user-avatar-spinner"
    color="grey"
    margin="4px"
    size="16px"
  />
</div>
`;

exports[`<UserAvatar /> should render component with no filename 1`] = `
<div
  className="user-avatar-container fit-parent"
>
  <div
    className="avatar-picture fit-parent"
  >
    <img
      alt="avatar"
      className="fit-parent"
      src="/images/avatar.png"
    />
  </div>
</div>
`;
